### ISSUE TITLE
How can I implement Async Call?

### ISSUE THREAD

**Issue by cwang92** [leangen/graphql-spqr#108]
I have a java Class name `FeedQuery`, inside that, I have:
```java
@GraphQLQuery(name = "feed")
@Override
public FeedController feed() {
    return feedController;
}
```

In the FeedController java class, I have several Apis such as:

```java
@GraphQLQuery(name = "m_ShipmentHeader")
  PagedElements<ShipmentLogicalRecord> getShipmentHeader(
      @GraphQLArgument(name = "filterArgs") Operator<ShipmentFilterArgs> filterArgs,
      @GraphQLArgument(name = "pagingArgs") ShipmentPagingArgs pagingArgs);

  @GraphQLQuery(name = "m_ShipmentLine")
  PagedElements<ShipmentLineRecord> getShipmentLine(
      @GraphQLArgument(name = "filterArgs") Operator<ShipmentLineFilterArgs> filterArgs,
      @GraphQLArgument(name = "pagingArgs") ShipmentLinePagingArgs pagingArgs);
```

So when I send query:
```graphql
{
       feed {
           m_ShipmentHeader(args....) {.....}
           m_ShipmentLine(args...){......}
       }
 }
```
Those two queries will do sequentially based on my understanding. How can I make them do parallel?

**Comment by kaqqao** [comment]
There's multiple ways to do this. The most direct is make the resolver functions (`getShipmentHeader` and `getShipmentLine` in your case) return `CompletableFuture`.
In that case you directly control when and how the resolution gets dispatched. E.g. if you create the `CompletableFuture`s via `CompletableFuture#supplyAsync`, they'll be resolved in parallel.

Another way is to use `ExecutorServiceExecutionStrategy` as the `queryStrategy` when constructing the `GraphQL` object. This strategy concurrently resolves the fields on the same level, but shouldn't be used carelessly as it also means even the trivial resolvers (simple getters) run on separate threads, which is rather needless.

For more info on `ExecutionStrategy` implementations, you may want to check out graphql-java samples, as this is really outside of SPQR.

**Comment by cwang92** [comment]
@kaqqao  Thanks for the quick reply.
Regarding the first way by doing CompletableFuture, Even though I return CompletableFuture of my resolver somehow is not working. I don't know whether I understand wrong or something.
I updated my method to:
```
  @GraphQLQuery(name = "m_ShipmentLine")
  CompletableFuture<PagedElements<GrnLineLogicalRecord>> getGrnLine(
      @GraphQLArgument(name = "filterArgs") Operator< ShipmentFilterArgs > filterArgs,
      @GraphQLArgument(name = "pagingArgs") ShipmentLinePagingArgs pagingArgs){
   return CompletableFuture.supplyAsync(my logic......)
}
note that these two resolver is under another class called feedQuery as I mentioned above, And I used that class to initialize the scheme. 
```
when I run, it seems that the graphql scheme can not be initialize. 

my schema code is looking like this:
```
@Inject
  private GraphQLSchemaGenerator graphQlSchemaGenerator;

  private GraphQL graphql;

  /**
   * Constructs the graphql schema once the QueryGraphQlSchema class is loaded.
   */
  @PostConstruct
  void init() {
    GraphQLSchema schema = graphQlSchemaGenerator
        .withAdditionalTypes(Lists.newArrayList(GraphQlDateType.DATE_TIME))
        .withOperationsFromSingleton(feedQuery)
        .withOperationsFromSingleton(aggregationQuery)
        .generate();
    graphql = GraphQL.newGraphQL(schema).build();
  }
```
and when I execute, I will do
```
ExecutionResult executionResult = graphql.execute(payload);
```
Somehow when I tried to generate the schema, it throw an exception saying ' Name must be non-null, non-empty and match [_A-Za-z][_0-9A-Za-z]* '

**Comment by cwang92** [comment]
@kaqqao  I'm using 0.9.3
Forget about the bug I told you now maybe. Am I doing the right thing for Aysnc ? If I change my resolver function return CompletableFuture type, will those function be executed parallel automatically and get the results?
Sorry I'm new to these things. Thanks in advance.

**Comment by kaqqao** [comment]
Oh, wow, that's an ancient version and `CompletableFuture` wasn't a thing back then! Please stay at the latest version at all times (0.9.7 at this moment) at least until 1.0 is released. I think the **async calls were first supported in 0.9.6**.

As for your question, if you return a `CompletableFuture` it is up to you to actually complete it. So you directly control which executions run in parallel. Using `CompletableFuture#supplyAsync` will immediately dispatch a thread from the common fork/join pool to execute  the computation asynchronously and complete the future. Graphql-java takes care of chaining the futures appropriately, so you don't need to worry about that.

In short, if you run field resolution asynchronously (as `supplyAsync` will do), yes, fields on the same level will get resolved in parallel as you desire, and yes your example seems correct.

If you call `GraphQL#executeAsync` instead of `GraphQL#execute` you'll even receive a `CompletableFuture<ExecutionResult>` instead of an `ExecutionResult` directly.

**Comment by cwang92** [comment]
@kaqqao Thank you.
I bump the version to 0.9.7, right now I'm seeing the 
Caused by: java.lang.ClassNotFoundException: graphql.schema.CoercingParseLiteralException
when I try generate the Graphql schema.  I did some search Found you have responded somewhere else saying it may be due to servlet.  But I don't have servlet dependency here.
What I only have is 
      <dependency>
        <groupId>com.graphql-java</groupId>
        <artifactId>graphql-java</artifactId>
        <version>3.0.0</version>
      </dependency>
Could it because the version of this one is also too old?

**Comment by cwang92** [comment]
@kaqqao Hey
I bumped the version of graphql-java to 8.0 and that error disappear.
Right not I'm having:
```
The type of member "getCrossEnterpriseOffsets" belonging to com.elementum.platform.query.models.schema.common.paging.PagedElements$PagingInfo is missing generic type parameters and can not be mapped. For details and possible solutions see https://github.com/leangen/graphql-spqr/wiki/Errors#ambiguous-member-type
```
I've taken look at the link you provided there, but not sure how to fix it now. Just curious how can  it working when I'm using version 0.9.3? Do you do some enhancement to check this error after bump the version? Do I have a way to avoid facing complaint these kinds of errors? Because my repo is an old and big repo, I think it will be much work there even though I fix this one.

**Comment by cwang92** [comment]
@kaqqao Thanks for the help!
Just to make sure, If I don't declare dependency for graphql-java, Can I still get it through the dependency on graphql-spqr?

As the error, I have:

```
@Data
public class PagedElements<E> {

  @NoArgsConstructor
  @Getter
  @Setter
  public static class PagingInfo {

    int totalCount;
    boolean hasMore;
    int offset;
    int count;

    List<CrossEnterpriseOffsets> crossEnterpriseOffsets;

  }
```
the method getCrossEnterpriseOffsets is from the lombok Anotation Getter. It's not used right now but will be used in the future. Don't understand why it's complaining this.

**Comment by kaqqao** [comment]
You unfortunately stumbled unto a very tricky edge case with generic type analysis. An inner static class enclosed by a generic owner.

I'll fix this for 0.9.8, but for the moment you have 2 options:

* Pull `PagingInfo` out, turning it into a top-level class (this will completely avoid the issue) **[recommended]**
* Use `withTypeTransformer(new DefaultTypeTransformer(true, true))` as **temporary** work-around until 0.9.8 is released

The first will avoid the issue and work correctly. The second will only shut the error up, but you'll still have to get the real fix for it before `crossEnterpriseOffsets` is fully usable (it will be treated as `Object` so you won't be able to select its sub-fields as you'd expect). 

>Just to make sure, If I don't declare dependency for graphql-java, Can I still get it through the dependency on graphql-spqr?
Yes, that's what dependency management is for.

**Comment by cwang92** [comment]
@kaqqao 
Thanks! that error is gone. Looking forward the 0.9.8 version.
Right now the GraphQLSchema is built successfully, but can not build the GraphQl object by the schema.
my code looks like:
```
  @Inject
  private FeedQuery feedQuery;

  @Inject
  private AggregationQuery aggregationQuery;

  @Inject
  private GraphQLSchemaGenerator graphQlSchemaGenerator;

  private GraphQL graphql;

  /**
   * Constructs the graphql schema once the QueryGraphQlSchema class is loaded.
   */
  @PostConstruct
  void init() {
    GraphQLSchema schema = graphQlSchemaGenerator
        .withAdditionalTypes(Lists.newArrayList(GraphQlDateType.DATE_TIME))
        .withOperationsFromSingleton(feedQuery)
        .withOperationsFromSingleton(aggregationQuery)
        .generate();
    graphql = GraphQL.newGraphQL(schema).build();
  }
}
```
And I got the error:

```
Caused by: java.lang.IllegalAccessError: tried to access method graphql.execution.ExecutionStrategy.<init>()V from class graphql.GraphQL$Builder
	at graphql.GraphQL$Builder.<init>(GraphQL.java:206)
	at graphql.GraphQL.newGraphQL(GraphQL.java:173)
	at com.elementum.platform.query.graphql.QueryGraphQlSchema.init(QueryGraphQlSchema.java:78)
	... 71 common frames omitted
```

**Comment by kaqqao** [comment]
As you can at e.g. [mvnrepository](https://mvnrepository.com/artifact/io.leangen.graphql/spqr/0.9.7), SPQR 0.9.7 depends on graphql-java 8.0. If you're seeing 2.0.0 (a very, very old version) you have some conflict on the classpath. Are you depending some other tool from the GraphQL eco system?
I suggest you do a detailed analysis and clear it up, as the exception you're showing is very likely caused the same problem.

If you're using IntelliJ IDEA Premium, it has an excellent dependency diagram. If not, use the regular Maven dependency tree and look for conflicts. Maybe your local repo is somehow corrupted? Try clearing everything under `io.leangen` and re-downloading. At any rate, SPQR has not referred to graphql-java 2.0.0 in a _very_ long time, so it's surely something messed up in your environment.

**Comment by cwang92** [comment]
I got duplicate classes under spqr when I run mvn clean install under my repo, what should I exclude here?
```
[WARNING] Found duplicate and different classes in [org.ow2.asm:asm-commons:6.0, org.ow2.asm:asm-tree:6.0, org.ow2.asm:asm:6.0]:
[WARNING]   module-info
```

**Comment by kaqqao** [comment]
Sorry, this is impossible to tell without knowing what else you have on your classpath and how is your project configured. Again, the issue at this point almost certainly has nothing to with SPQR. SPQR depends on asm 6.0 (will be 6.1 in SPQR 0.9.8) and it by itself can't duplicate anything. You might have something else depending on asm, or maybe you have multiple SPQR versions on your classpath or another conflicting library. I simply can't tell from here.

Debugging your project setup is not something I can do over messages, nor is in the scope of this project. The current discussion no longer has anything to do with the original question (which was about parallel resolution of fields). You'll really have to troubleshoot further on your own as you're the only one with an insight into your environment.

**Comment by cwang92** [comment]
Yes. Thanks very much for the help
